### DISCUSSION TITLE
Stacked Git Internals Discussion

### DISCUSSION THREAD

**Discussion by NonLogicalDev** [stacked-git/stgit#167] [Ideas]
@jpgrayson I was looking into Stacked Git's Stack Implementation in relation to #163, and found myself a little confused about the 
reasons behind the `patches` directory on the `stacks/<branch>` branch.

What are those files used for? I found the logic to write them:
* https://github.com/stacked-git/stgit/blob/master/stgit/lib/log.py#L265

But could not find where if it is ever actually read back. It seems they largely have the same information as the commit objects.

```
Bottom: $(git rev-parse "$PATCH_OID^1^{tree}")
Top: $(git rev-parse "$PATCH_OID^{tree}")
Author/Date: $(git cat-file -p $PATCH_OID | grep author)
```

Are those files still used?

After experimenting with a shell script that fixes up the topmost commit it seems like patches get regenerated anyway, even if incorrectly updated by the script:

https://gist.github.com/NonLogicalDev/bd110dc0a9c1519f6d79a4138b3c80de

**Comment by jpgrayson** [comment]
You are correct that those per-patch metadata files are not parsed/used by StGit directly.

Their purpose is to provide rich patch logging. This is a fairly deeply buried feature--you would only encounter the content of these patch metadata files when running `stg log --diff <patchname>`.

This feature has been there since before I became involved in the project. I've preserved this feature over the ages, but I've only occasionally found its utility useful. I may not appreciate the workflows or use cases where this is really helpful.

> it seems like patches get regenerated anyway, even if incorrectly updated by the script

Are you referring to the patch refs? I.e. `refs/patches/<branch>/<patchname>`? An issue StGit has to deal with is that patch refs are redundantly found both in `refs/patches/<branch>` and in `refs/stacks/<branch>:stack.json`. The problem is that there is opportunity either via StGit bugs or the user doing manual ref updates for patch refs in `refs/patches/<branch>` to be out-of-sync with `stack.json`. The way StGit resolves this is to treat `stack.json` as canon and synchronize `refs/patches/<branch>` whenever a `Stack` object is initialized; i.e. resynchronize everytime `stg` is run.

Alternatively, if you are referring to the patch metadata files in `refs/stacks/<branch>:patches/`, yes, that tree is regenerated whenever any change is made to the patch series.
